### ISSUE TITLE
Media provider Uri permission problem

### ISSUE THREAD

**Issue by secu2** [gotev/android-upload-service#269]
### Version info
* Android Upload Service version: 3.2.3
* Android version and API version: 19+ & targetSdkVersion 25
* HTTP stack (e.g. HurlStack or OkHttpStack): OkHttpStack

### What did you expect?
The upload service to be able to read a content://com.android.providers.media.documents/ Uri

### What happened instead?
A permission problem when trying to query the Cursor :  [Caused by](https://github.com/gotev/android-upload-service/blob/0f753eab80d7d9d1fb7a1e8e9bc8039f2c212cda/uploadservice/src/main/java/net/gotev/uploadservice/schemehandlers/ContentSchemeHandler.java#L54)
```
java.lang.SecurityException: Permission Denial: reading com.android.providers.media.MediaDocumentsProvider uri content://com.android.providers.media.documents/document/image:47 from pid=14195, uid=10076 requires android.permission.MANAGE_DOCUMENTS, or grantUriPermission()
```

### Library initialization code:
```java
public BinaryUploadRequest setFileToUpload(final Uri file, final String fileName) throws FileNotFoundException {
        this.fileName = fileName;

        /* ------- DEBUG -------
        Cursor cursor = context.getContentResolver().query(file, null, null, null, null);
        if (cursor == null) {
            Logger.error(getClass().getSimpleName(), "null cursor for " + file + ", returning size 0");
        }
        int sizeIndex = cursor.getColumnIndex(OpenableColumns.SIZE);
        cursor.moveToFirst();
        long size = cursor.getLong(sizeIndex);
        cursor.close();
        Log.d(TAG, "setFileToUpload: Successfully retrieved file size: "+size);
        /* ----- END DEBUG ----- */

        super.context.grantUriPermission("net.gotev.uploadservice", file, Intent.FLAG_GRANT_READ_URI_PERMISSION);

        String fullUrl = serverUrl+path+fileName;
        Log.d(TAG, "Updating URL to "+fullUrl);
        this.params.setServerUrl(fullUrl);
        realm.executeTransaction(...);

        return super.setFileToUpload(Uri.decode(file.toString()));
```
- [Commented code] this code is causing the problem when used in the runnable part of the upload service but works here as my app has all the needed permissions to read a content Uri.
- As you can see I tried to grantUriPermission() in vain

### Request code:
```java
MyCustomBinaryUploadRequest(getContext(), UUID.randomUUID(), currentUserInstance.serverURL, currentUserInstance.currentAbsoluteDirPath(), null)
.setFileToUpload(fileUri, fileName)
.startUpload();
```
#### Where have you added the request code?
- [x] Activity
- [ ] Service
- [ ] Other class (add additional info about it)

### LogCat output (please [set log level to DEBUG](https://github.com/gotev/android-upload-service/wiki/Recipes#logging-) first)
```
D/MyCustomBinaryUploadRequest: Upload f29f61c7-0eaf-4ebe-a0dd-137171bc37e8 registered in DB for user drive5650
D/MyCustomBinaryUploadRequest: Updating URL to https://drive5650.phpnetstorage.eu/IMG_20170518_163139.jpg
D/MyCustomBinaryUploadRequest: Registering file to upload for request f29f61c7-0eaf-4ebe-a0dd-137171bc37e8 in DB for user drive5650 (File path on device : content://com.android.providers.media.documents/document/image%3A47 Remote path : /)
D/MyCustomBinaryUploadRequest: Registered file to upload for request f29f61c7-0eaf-4ebe-a0dd-137171bc37e8) in DB for user drive5650
D/AcceuilActivity: Launched upload of file content://com.android.providers.media.documents/document/image:47 to https://drive5650.phpnetstorage.eu/. (UploadID : f29f61c7-0eaf-4ebe-a0dd-137171bc37e8)
D/AcceuilActivity: Registering the uploadservice status broadcast listener.
I/UploadService: UploadService - Starting service with namespace: net.gotev, upload pool size: 2, 1s idle thread keep alive time. Foreground execution is enabled
D/UploadService: UploadService - Successfully created new task with class: net.gotev.uploadservice.BinaryUploadTask
D/UploadService: UploadService - f29f61c7-0eaf-4ebe-a0dd-137171bc37e8 now holds the foreground notification
D/UploadService: HttpUploadTask - Starting upload task with ID f29f61c7-0eaf-4ebe-a0dd-137171bc37e8
I/UploadService: UploadTask - Broadcasting error for upload with ID: f29f61c7-0eaf-4ebe-a0dd-137171bc37e8. Permission Denial: reading com.android.providers.media.MediaDocumentsProvider uri content://com.android.providers.media.documents/document/image:47 from pid=14195, uid=10076 requires android.permission.MANAGE_DOCUMENTS, or grantUriPermission()

```
### Server side
Server side is an Apache 2 (WebDav)

### More info
- I also tried in vain adding permissions to the uploadservice module ```<uses-permission android:name="android.permission.MANAGE_DOCUMENTS" />```

**Comment by gotev** [comment]
You have to simply add the required permission to your app and request it
runtime for android > 6.0 before calling the service. Permission handling
is not a library concern

Il 19 mag 2017 6:21 PM, "secu2" <notifications@github.com> ha scritto:

**Comment by secu2** [comment]
My app has all the permissions requested:
```
<uses-permission android:name="android.permission.MANAGE_DOCUMENTS" />
<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
[...]
```

I even retrieve the file name with no problem before starting the service : 
```
 Cursor cursor = getActivity().getContentResolver().query(fileUri, null, null, null, null, null);
                    if(cursor != null && cursor.moveToFirst()){
                        String fileName = cursor.getString(cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME));
}
```

**Comment by secu2** [comment]
Seems like using an encoded Uri worked for me :
`[BinaryUploadRequest].setFileToUpload(file.toString());`

instead of :
`[BinaryUploadRequest].setFileToUpload(Uri.decode(file.toString()));`

**Comment by gajicm93** [comment]
I'm having the exact same problem. And using the encoded Uri doesn't fix the problem for me? Please help? @secu2 @gotev

**Comment by gajicm93** [comment]
Fixed it!

The problem is in with the Intent you pick up content with, and whether you properly take read permission for Uri. Here is the solution if anybody needs it:

https://stackoverflow.com/questions/25414352/how-to-persist-permission-in-android-api-19-kitkat/29588566#29588566

**Comment by gotev** [comment]
@gajicm93 Great!
